app/components-react/root/GLVolmeters.tsx

```diff
@@ -0,0 +1,438 @@
+      this.frameNumber = -1;
+      this.firstFrameTime = now;
+    }
+
+    const timeElapsed = now - this.firstFrameTime;
+    const timeBetweenFrames = 1000 / this.fpsLimit;
+    const currentFrameNumber = Math.ceil(timeElapsed / timeBetweenFrames);
+
+    if (currentFrameNumber !== this.frameNumber) {
+      // it's time to render next frame
+      this.frameNumber = currentFrameNumber;
+      // don't render sources then channelsCount is 0
+      // happens when the browser source stops playing audio
+      this.drawVolmeters();
+    }
+    this.requestedFrameId = requestAnimationFrame(t => this.onRequestAnimationFrameHandler(t));
+  }
+
+  private initWebglRendering() {
+    const vShader = getDefined(compileShader(this.gl, vShaderSrc, this.gl.VERTEX_SHADER));
+    const fShader = getDefined(compileShader(this.gl, fShaderSrc, this.gl.FRAGMENT_SHADER));
+    this.program = getDefined(createProgram(this.gl, vShader, fShader));
+
+    const positionBuffer = this.gl.createBuffer();
+    this.gl.bindBuffer(this.gl.ARRAY_BUFFER, positionBuffer);
+
+    // Vertex geometry for a unit square
+    // eslint-disable-next-line
+    const positions = [
+      0, 0,
+      0, 1,
+      1, 0,
+      1, 0,
+      0, 1,
+      1, 1,
+    ];
+
+    this.gl.bufferData(this.gl.ARRAY_BUFFER, new Float32Array(positions), this.gl.STATIC_DRAW);
+
+    // look up where the vertex data needs to go.
+    this.positionLocation = this.gl.getAttribLocation(this.program, 'a_position');
+
+    // lookup uniforms
+    this.resolutionLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_resolution'));
+    this.translationLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_translation'));
+    this.scaleLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_scale'));
+    this.volumeLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_volume'));
+    this.peakHoldLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_peakHold'));
+    this.bgMultiplierLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_bgMultiplier'));
+
+    this.gl.useProgram(this.program);
+
+    const warningLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_warning'));
+    this.gl.uniform1f(warningLocation, this.dbToUnitScalar(WARNING_LEVEL));
+
+    const dangerLocation = getDefined(this.gl.getUniformLocation(this.program, 'u_danger'));
+    this.gl.uniform1f(dangerLocation, this.dbToUnitScalar(DANGER_LEVEL));
+
+    // Set colors
+    this.setColorUniform('u_green', GREEN);
+    this.setColorUniform('u_yellow', YELLOW);
+    this.setColorUniform('u_red', RED);
+  }
+
+  private setColorUniform(uniform: string, color: number[]) {
+    const location = this.gl.getUniformLocation(this.program, uniform);
+    // eslint-disable-next-line
+    this.gl.uniform3fv(location, color.map(c => c / 255));
+  }
+
+  private setCanvasSize() {
+    const $parent = getDefined(this.$refs.canvas?.parentElement);
```

Comment:
not sure if i understand the interaction of `getDefined` and safe navigation here

Reply:
In theory, using `getDefined` here produces a better stack trace, and, if we see it in sentry it should be prioritized because it means we made an incorrect assumption about whether something could be null.

In practice - I'm not sure we're very diligent about fixing these when we make incorrect assumptions.

From a runtime perspective it doesn't matter - it's runtime exception either way.